### PR TITLE
Add OOM section to Best Practices

### PR DESCRIPTION
**Pull request by sarahyurick**

### CODE REVIEW

**Review by ryantwolf — CHANGES_REQUESTED** [review]
I left a couple of suggestions. Let me know what you think.

File: docs/user-guide/bestpractices.rst [hunk 1]
```diff
@@ -16,4 +16,32 @@ Here are the methods in increasing order of compute required for them.
...
+
+Add More GPUs
+~~~~~~~~~~~~~
+If possible, scale your system by adding more GPUs.
+This provides additional VRAM (Video Random Access Memory), which is crucial for holding datasets and intermediate computations.
+Thus, adding more GPUs allows you to distribute the workload, reducing the memory load on each GPU.
+
+Utilize RMM Options
+~~~~~~~~~~~~~~~~~~~
+`RAPIDS Memory Manager (RMM) <https://github.com/rapidsai/rmm>`_ is a package that enables you to allocate device memory in a highly configurable way.
```
[the remark this answers is not in the document]
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst:35]: I believe these flags should mainly come in handy for fuzzy deduplication, as the initial bug raised used them to solve OOM issues with the connected components step.

cc @ayushdg if this seems right to you?

File: docs/user-guide/bestpractices.rst [hunk 2]
```diff
@@ -16,4 +16,32 @@ Here are the methods in increasing order of compute required for them.
...
+Here are some features which can help optimize memory usage:
+
+* Enable asynchronous memory allocation: Use the ``--rmm-async`` flag to allow RMM to handle memory allocation more efficiently, by allocating and deallocating GPU memory asynchronously.
```
[the remark this answers is not in the document]
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst:38]: It looks like we hardcode `rmm_async` in some cases (see [here](https://github.com/NVIDIA/NeMo-Curator/blob/main/nemo_curator/utils/distributed_utils.py#L72)). Should we make it configurable by the user?

It looks like there isn't any way to configure the `rmm_release_threshold` in the Python API. Should we create an issue for this?

File: docs/user-guide/bestpractices.rst [hunk 3]
```diff
@@ -16,4 +16,32 @@ Here are the methods in increasing order of compute required for them.
...
+* Set a memory release threshold: For example, ``--rmm-release-threshold 50GB`` can help prevent holding onto excess memory, releasing unused memory when a certain limit is reached. Please keep in mind that using this flag may degrade performance slightly as RMM is busy releasing the unused memory.
+
+Fuzzy Deduplication Guidelines
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Fuzzy deduplication is one of the most computationally expensive algorithms within the NeMo Curator pipeline.
+Here are some suggestions for managing memory use during fuzzy deduplication:
+
+* Reduce bucket counts: During deduplication, the data is grouped into buckets to compare and identify near-duplicate documents. Reducing the number of buckets can help decrease the number of data points loaded into memory for comparison. However, a smaller bucket count can reduce the accuracy of deduplication, so it is important to find an optimal balance between memory usage and deduplication accuracy. You can experiment with this by using the ``buckets_per_shuffle`` parameter when initializing your ``FuzzyDuplicatesConfig``.
```
[the remark this answers is not in the document]
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst]: Updated, thanks!

**Comment by sarahyurick** [comment]
Thanks @ryantwolf ! Updated for whenever you get a chance to review again.

File: docs/user-guide/bestpractices.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryantwolf) [reply docs/user-guide/bestpractices.rst:33]: So long as they are potentially useful for the user to configure, we should add them as options to the Python API and CLI. Keep their defaults at what they are in `LocalCUDACluster`.

**Comment by ryantwolf** [comment]
@sarahyurick I left a few more comments.

File: docs/user-guide/bestpractices.rst [hunk 4]
```diff
@@ -16,4 +16,40 @@ Here are the methods in increasing order of compute required for them.
...
+* Reduce buckets per shuffle: Because duplicates are still considered bucket by bucket, reducing the ``buckets_per_shuffle`` parameter in the ``FuzzyDuplicatesConfig`` does not affect accuracy. Instead, reducing the buckets per shuffle helps lower the amount of data being transferred between GPUs. However, using a lower ``buckets_per_shuffle`` will increase the time it takes to process the data.
+* Adjust files per partition: Processing large datasets in smaller chunks can help reduce the memory load. When reading data with ``DocumentDataset.read_json`` or ``DocumentDataset.read_parquet``, start with a smaller ``files_per_partition`` value and increase as needed.
```
Comment (ayushdg) [inline docs/user-guide/bestpractices.rst:49]: We can also make the recommendation to target no larger than `2GB` partitions for the time being. As a todo: Some of the suggestions made here: https://github.com/rapidsai/cudf/pull/16821 may also apply to us. We can link to it in the future if needed.

File: docs/user-guide/bestpractices.rst [hunk 5]
```diff
@@ -16,4 +16,40 @@ Here are the methods in increasing order of compute required for them.
...
+
+Add More GPUs
+~~~~~~~~~~~~~
+If possible, scale your system by adding more GPUs.
+This provides additional VRAM (Video Random Access Memory), which is crucial for holding datasets and intermediate computations.
+Thus, adding more GPUs allows you to distribute the workload, reducing the memory load on each GPU.
```
Comment (ayushdg) [inline docs/user-guide/bestpractices.rst:55]: Todo for the future: Share rough dataset sizes and how many GPUs we've been able to run this on internally. That can give a sense of the requirements.

Comment (miguelusque) [reply docs/user-guide/bestpractices.rst:55]: Would it be possible to also mention the parameters we can use to reduce out_of_memory in CPUs, such as **--n-workers** or **--batch-size**?

**Comment by sarahyurick** [comment]
TODO: How to report/capture GPU memory and utilization for a specific step.

References:

- `from dask.distributed import performance_report` excludes GPU information
- Track RMM memory pool: https://github.com/rapidsai/cugraph/blob/8b1a437275b2a0c4477afec4cc41eb3281b12517/python/cugraph/cugraph/testing/mg_utils.py#L324-L345
- RMM docs: https://docs.rapids.ai/api/rmm/stable/guide/#memory-statistics-and-profiling
- Dashboard with instant RMM memory utilization: https://github.com/dask/distributed/pull/7718

File: docs/user-guide/bestpractices.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst:33]: Opened https://github.com/NVIDIA/NeMo-Curator/issues/260 to track this.

File: docs/user-guide/bestpractices.rst [hunk 6]
```diff
@@ -16,4 +16,40 @@ Here are the methods in increasing order of compute required for them.
...
+.. code-block:: bash
+
+  python your_script.py --rmm-async --rmm-release-threshold 50GB
+
+Fuzzy Deduplication Guidelines
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Fuzzy deduplication is one of the most computationally expensive algorithms within the NeMo Curator pipeline.
+Here are some suggestions for managing memory use during fuzzy deduplication:
+
+* Reduce bucket counts: During deduplication, the data is grouped into buckets to compare and identify near-duplicate documents. Increasing the number of buckets increases the probability that two documents within a given Jaccard similarity score are marked as duplicates. This leads to more accurate results as it reduces the number of false negatives. However, increasing the number of buckets also increases the memory requirements from the increased number of hashes it needs to store. Thus, it is important to find an optimal balance between memory usage and deduplication accuracy. You can experiment with this by using the ``num_buckets`` parameter when initializing your ``FuzzyDuplicatesConfig``.
```
[the remark this answers is not in the document]
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst:47]: Updating now, LMK what you think.

File: docs/user-guide/bestpractices.rst (hunk 4 above)
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst:49]: Updating now and adding link.

File: docs/user-guide/bestpractices.rst (hunk 5 above)
Comment (sarahyurick) [reply docs/user-guide/bestpractices.rst:55]: @ayushdg I can add a TODO comment for now.

@miguelusque right now the goal of this PR is to help with GPU OOM errors. I can add a general note to ensure the user is aware of the different parameters they can consider when calling `get_client`. Hopefully this aligns with your request?

File: docs/user-guide/bestpractices.rst [hunk 7]
```diff
@@ -16,4 +16,64 @@ Here are the methods in increasing order of compute required for them.
...
+If possible, scale your system by adding more GPUs.
+This provides additional VRAM (Video Random Access Memory), which is crucial for holding datasets and intermediate computations.
+Thus, adding more GPUs allows you to distribute the workload, reducing the memory load on each GPU.
+
+.. TODO: Share rough dataset sizes and how many GPUs we've been able to run this on internally; that can give a sense of the requirements.
+
+Report GPU Memory and Utilization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When debugging your GPU memory errors, it can be useful to capture and understand your GPU usage per step in the NeMo Curator pipeline. For this, using RMM's `Memory stastics and profiling <https://docs.rapids.ai/api/rmm/stable/guide/#memory-statistics-and-profiling>`_ can be helpful. You may also refer to `this article <https://medium.com/rapids-ai/monitoring-dask-rapids-with-prometheus-grafana-96eaf6b8f3a0>`_, for a general tutorial including how to monitor GPUs with a dashboard.
```
Comment (sarahyurick) [inline docs/user-guide/bestpractices.rst:79]: Hey @VibhuJawa I was wondering what you think of this?

File: docs/user-guide/bestpractices.rst [hunk 8]
```diff
@@ -16,4 +16,64 @@ Here are the methods in increasing order of compute required for them.
...
```
Comment (ronjer30) [inline docs/user-guide/bestpractices.rst:79]: The Dask Diagnostics [Dashboard](https://docs.dask.org/en/stable/dashboard.html) might be a good starting point to view GPU utilization and memory at a high level.  Additionally, an example of how to use the RMM profiler specifically with a NeMo Curator step would be useful.

File: docs/user-guide/bestpractices.rst (hunk 5 above)
Comment (miguelusque) [reply docs/user-guide/bestpractices.rst:72]: Hi @sarahyurick , my bad. I think some of those parameters such as batch size and num_files are useful for both COU and GPU clusters. I hope it helps!

P.S.: Happy to address them in a different PR. Thanks!!!

File: docs/user-guide/bestpractices.rst [hunk 9]
```diff
@@ -16,4 +16,66 @@ Here are the methods in increasing order of compute required for them.
...
+    * For other suggestions on best practices regarding reading data with Dask, please refer to `Dask cuDF Best Practices <https://github.com/rapidsai/cudf/blob/branch-24.10/docs/dask_cudf/source/best_practices.rst#reading-data>`_.
+
+Using the ``get_client`` Function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For both GPU and CPU operations, we provide the ``get_client`` to initialize your Dask client with a ``LocalCUDACluster`` or ``LocalCluster``, respectively. While the NeMo Curator team has established default values for the parameters of the ``get_client`` function that are suitable for most scenarios, it is useful to understand these parameters and become familiar with them to ensure optimal performance and adherence to best practices when working with Dask configurations and setups.
+
+Please refer to the `distributed_utils.py <https://github.com/NVIDIA/NeMo-Curator/blob/main/nemo_curator/utils/distributed_utils.py>`_ script for detailed docstrings, especially for the ``get_client`` function, and the ``start_dask_gpu_local_cluster`` and ``start_dask_cpu_local_cluster`` functions which are called by ``get_client``.
```
Comment (ryantwolf) [inline docs/user-guide/bestpractices.rst:66]: I'd point users to the API docs for them: https://docs.nvidia.com/nemo-framework/user-guide/latest/datacuration/api/dask.html. You can keep the exact location in the code too, but the API docs should be the first recommendation

File: docs/user-guide/bestpractices.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryantwolf) [reply docs/user-guide/bestpractices.rst:55]: Shoot somehow a comment I made earlier about this didn't get published. Here is what the comment was:

Nit: Remove the sentence "This leads to more accurate results as it reduces the number of false negatives". I should have been clearer in my previous comment, changing the number of buckets impacts the results of the algorithm. It impacts the accuracy, but it's subjective whether it increases the accuracy or lowers the accuracy. I don't think it's increasing accuracy or removing false negatives, it's more just removing negatives in general. It lowers the Jaccard threshold that is considered a duplicate.

**Review by ryantwolf — COMMENTED** [review]
I have one more minor thing and I review comment that somehow didn't get published before, sorry about that.

**Comment by sarahyurick** [comment]
Thanks @ryantwolf ! Updated.

**Review by ryantwolf — APPROVED** [review]
Good on my end, thank you!

**Pull request merged**